File: solutions_11-20/Euler_14.py
# -*- coding: utf-8 -*-

'''
The following iterative sequence is defined for the set of positive integers:

n → n/2 (n is even)
n → 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:
13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1

It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.
'''

import timeit
import operator


start = timeit.default_timer()

def collatz(n):
    var = []
    while n != 1:
        if n % 2 != 0:
            n = (n * 3) + 1
            var.append(n)
        if n % 2 == 0:
            n = n / 2 
            var.append(n)
    return len(var)


def euler_14():
    data = map(collatz,range(1,1000000))
    enum_data = list(enumerate(data, 1))
    enum_data.sort(key=operator.itemgetter(1))
    return enum_data[len(enum_data)-1]  


print "Answer: %s" % euler_14()[0]
stop = timeit.default_timer()
print "Time: %s" % str(stop - start)

File: Java/java_solutions_11-20/Euler_14.java
/*
/   Project Euler problem #14 Longest Collatz sequence
/
/   Which starting number, under one million, produces the longest chain?
/
*/


public class Euler_14
{
    public static long collatzSequence(long chain)
    {
        long sequenceCount = 0;
        while (chain > 1) {
            if (chain % 2 == 0) {
                chain /= 2;
            } else {
                chain = (chain * 3) + 1;
            }
            sequenceCount++;
        }
        return sequenceCount;
    }

    public static long findLongestChain(long limit, 
                                       long candidate, 
                                       long longestChain, 
                                       long longestChainStart) {
        if (candidate > limit) {
            return longestChainStart; 
        }

        long currentChain = collatzSequence(candidate);
        if (currentChain > longestChain) {
            longestChain = currentChain;
            longestChainStart = candidate;
        }

        return findLongestChain(limit, candidate + 1, longestChain, 
                                                      longestChainStart);
    }

    public static void main(String[] args)
    {
        long start = System.nanoTime();
        long answer = findLongestChain(1000000, 0, 0, 0);
        long stop = System.nanoTime();
        System.out.println("Answer: " + answer); 
        System.out.printf("Time: %.4f\n", ((float) stop - start) / 1_000_000_000);            
    }
}
